### ISSUE TITLE
SA1009,1111,1114,1115 trigger when using compiler directives in method signatures

### ISSUE THREAD

**Issue by jessehouwing** [DotNetAnalyzers/StyleCopAnalyzers#1623]
Unfortunately due to Microsoft renaming a bunch of interfaces, we had to use compiler directives to pick the right interface for the version being compiled.

It would be great if Stylecop would ignore compiler directive lines and not treat them as whitespace.

````
#pragma warning disable SA1111, SA1115, SA1114, SA1009
        public EventNotificationStatus ProcessEvent(
#if TFS20151
            IVssRequestContext requestContext,
#else
            TeamFoundationRequestContext context,
#endif
            NotificationType notificationType,
            object notificationEventArgs,
            out int statusCode,
            out string statusMessage,
            out ExceptionPropertyCollection properties)
#pragma warning restore SA1111, SA1115, SA1114, SA1009
        {
````

And:

```
        private Uri GetCollectionUriFromContext(
+#if TFS20151
+            IVssRequestContext requestContext
+#else
+            TeamFoundationRequestContext requestContext
+#endif
+            )
         {
```

**Comment by vweijsters** [comment]
@jessehouwing Thanks for reporting this issue.

I agree that said rules should not report diagnostics in this case.

**Comment by vweijsters** [comment]
Grabbing this

**Comment by vweijsters** [comment]
@jessehouwing why did you disable SA1009? Its seems to me that there would be no need for that.

The only reason I can think of is:
```csharp
    public void TestMethod(
#if SOMEEVENT
        IInterface1 instance
#else
        IInterface2 instance
#endif
        )
    {
    }
```

but that construction is incorrect in my opinion and should be written as:
```csharp
    public void TestMethod(
#if SOMEEVENT
        IInterface1 instance)
#else
        IInterface2 instance)
#endif
    {
    }
```

**Comment by vweijsters** [comment]
The observation above probably also holds true for SA1111.

**Comment by jessehouwing** [comment]
That would work, though I prefer to have the the least amount possible between my compiler directives. Adding a new parameter at the end would require me to update both compiler directive parts.

I can see how your layout would work as well.

I ended up finding a better solution anyway, using:

```
#if TFS2015u1
using IVssRequestContext = Microsoft.TeamFoundation.Framework.Server.IVssRequestContext;
#else
using IVssRequestContext = Microsoft.TeamFoundation.Framework.Server.TeamFoundationRequestContext;
#endif
```

Because the interface was simply renamed, that results in cleaner code overall.

**Comment by vweijsters** [comment]
OK, thanks for the feedback. Your new solution is preferable indeed :+1:

I'll still fix this for SA1114 and SA1115, since that still is a valid use case.

**Comment by jessehouwing** [comment]
Thanks!
